llvm/lib/Target/NVPTX/SYCL/GlobalOffset.cpp

```diff
@@ -0,0 +1,368 @@
+//===--------- GlobalOffset.cpp - Global Offset Support for CUDA --------- ===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass operates on SYCL kernels being compiled to CUDA. It looks for uses
+// of the `llvm.nvvm.implicit.offset` intrinsic and replaces it with a offset
+// parameter which will be threaded through from the kernel entry point.
+//
+//===----------------------------------------------------------------------===//
+
+#include "GlobalOffset.h"
+
+#include "../MCTargetDesc/NVPTXBaseInfo.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsNVPTX.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Target/TargetIntrinsicInfo.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "globaloffset"
+
+namespace llvm {
+void initializeGlobalOffsetPass(PassRegistry &);
+} // end namespace llvm
+
+namespace {
+
+class GlobalOffset : public ModulePass {
+public:
+  static char ID;
+  GlobalOffset() : ModulePass(ID) {}
+
+  bool runOnModule(Module &M) override {
+    if (skipModule(M))
+      return false;
+
+    llvm::Function *ImplicitOffsetIntrinsic =
+        M.getFunction(Intrinsic::getName(Intrinsic::nvvm_implicit_offset));
+    KernelImplicitArgumentType =
+        ArrayType::get(Type::getInt32Ty(M.getContext()), 3);
+    ImplicitOffsetPtrType = Type::getInt32Ty(M.getContext())->getPointerTo();
+    assert(
+        (!ImplicitOffsetIntrinsic ||
+         ImplicitOffsetIntrinsic->getReturnType() == ImplicitOffsetPtrType) &&
+        "Intrinsic::nvvm_implicit_offset does not return the expected "
+        "type");
+
+    if (!ImplicitOffsetIntrinsic || ImplicitOffsetIntrinsic->use_empty()) {
```

Comment:
It would be more readable if we move this early exit to line 48 - right after `ImplicitOffsetIntrinsic` is defined.

Reply:
Good call. I've moved this up.